mtt/src/mtt_types.rs

```diff
@@ -94,27 +93,512 @@ impl MttL2EntryType {
+    // The MTT update is skipped if the page is already contained in a spanning
+    // range of the same type.
+    // The desired page size is assumed to be 4K, and larger spans are broken
+    // down into smaller regions.
+    fn set_page_range_memory_type(&mut self, memory_type: MttMemoryType) -> Result<bool> {
+        self.validate_page_range()?;
+        // At this point, success is guaranteed because we have verified that all L2
+        // entries are valid, and that there are sufficient backing L1 pages if needed.
+        let mut mtt_needs_invalidation = false;
+        let page_iter = self
+            .current_phys_addr
+            .iter_from_with_size(self.page_size)
+            .ok_or(InvalidAddress)?;
+        for phys_addr in page_iter.take(self.num_pages) {
+            self.current_phys_addr = phys_addr;
+            self.current_l2_entry = Self::get_mtt_l2_entry(phys_addr, self.inner.mtt_l2_base)?;
+            let is_confidential = matches!(self.get_memory_type()?, MttMemoryType::Confidential);
+            // If the address is already spanned by a range of the same memory type
+            // there's nothing to do. If the memory types differ, we may have to split
+            // the spanning range into smaller regions.
+            if (memory_type == MttMemoryType::Confidential && !is_confidential)
+                || (memory_type == MttMemoryType::NonConfidential && is_confidential)
+            {
+                mtt_needs_invalidation = true;
+                match self.get_l2_entry_type() {
+                    MttConfidential1G | MttNonConfidential1G => {
+                        self.convert_1gb_to_64mb_range()?;
+                        self.convert_64mb_to_l1_dir()?;
+                        self.set_4k_page_memory_type(memory_type)?;
+                    }
+                    Mtt2MBPages(_) => {
+                        self.convert_64mb_to_l1_dir()?;
+                        self.set_4k_page_memory_type(memory_type)?;
+                    }
+                    MttL1Dir(_) => {
+                        self.set_4k_page_memory_type(memory_type)?;
+                    }
+                }
+            }
+        }
+        Ok(mtt_needs_invalidation)
+    }
+}
+
+/// Sets the type of the page range starting with `phys_addr` to the type specified
+/// in `memory_type`. `PageSize` specifies the size of the page (4K, 2MB, 512MB, 1GB), and
+/// `num_pages` is the number of pages in the range in terms of `PageSize`. Note that at
+/// present, only 4K pages are supported.
+/// On success, the function returns a boolean value to indicate whether the MTT needs to
+/// be invalidated. If the return value is true, the call must be followed by a subsequent
+/// upstream call to invalidate the MTT and ensurethat the changes take effect in hardware.
+// TODO: Add support for larger pages types.
+pub fn set_page_range_memory_type(
+    phys_addr: SupervisorPageAddr,
+    page_size: PageSize,
+    num_pages: usize,
```

Comment:
So how would we end up using `page_size` here? It's not part of the host-facing API, so that means Salus has to pick the appropriate size before calling this function. But `set_page_range_memory_type()` is already doing the splitting down to 4k automatically -- can't we add logic to use 2M or 1G when appropriate, instead of relying on higher-level logic to figure it out?